File: packages/provide-uterm-client/src/provide/uterm/ai/patterns.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""Structural ReDoS guard for untrusted (LLM/user-supplied) regex patterns.

The MCP tool surface lets a caller — including a low-privilege ``viewer`` role
via ``session_watch`` / ``session_subscribe`` — hand us a regex that we compile
with stdlib :mod:`re`.  A pure length cap does not bound catastrophic
backtracking for *short* pathological patterns such as ``(a+)+$``, so before any
``re.compile`` we reject the cheap, well-known catastrophic constructs:

* **nested quantifiers** — a quantified group whose body is itself quantified
  (``(a+)+``, ``(a*)*``, ``(a+)*``, ``(a*)+``, ``(\\w+)+`` …), the classic
  exponential-backtracking shape; and
* **quantified backreferences** — ``\\1+``, ``(\\1)+`` …, which also drive
  pathological backtracking.

Neither stdlib :mod:`re` (no time bound) nor a new third-party engine
(``re2``/``regex`` are not project dependencies) is used: this is a structural
denylist applied *in addition to* the length cap.

Residual risk (documented honestly): this does NOT catch every ReDoS shape —
notably overlapping alternations like ``(a|a)*`` or quantifier interactions
across adjacent tokens — because detecting those structurally without large
false-positive collateral is hard.  It removes the cheap, classic amplification
paths an LLM is most likely to emit; egress/time bounds at the server remain the
defence-in-depth backstop.
"""

from __future__ import annotations

import re

# A backreference token: ``\1`` .. ``\99`` (stdlib ``re`` allows up to 99 groups).
_BACKREF = re.compile(r"\\[1-9][0-9]?$")

# A quantified backreference *anywhere* in the pattern: ``\1+``, ``\2*``,
# ``\3{2,5}`` — the ``\`` is not itself escaped because the scan below skips
# escaped pairs when collecting groups, but a top-level backref+quantifier needs
# its own check (it is not wrapped in a group).
_QUANTIFIED_BACKREF = re.compile(r"\\[1-9][0-9]?[+*{]")

# Quantifier characters that, when they immediately follow a group close paren,
# make that group "repeated".
_QUANTIFIER_OPENERS = frozenset("+*{")


def _group_bodies_with_following_char(pattern: str) -> list[tuple[str, str]]:
    """Return ``(body, char_after_close)`` for every balanced ``(...)`` group.

    Escaped parens (``\\(`` / ``\\)``) are treated as literals, not group
    delimiters.  Unbalanced closes are ignored (the downstream stdlib compiler
    rejects the pattern as invalid), so this never raises on malformed input.
    """
    stack: list[int] = []
    results: list[tuple[str, str]] = []
    i = 0
    n = len(pattern)
    while i < n:
        char = pattern[i]
        if char == "\\":
            # Skip the escaped character so ``\(`` / ``\)`` are literals.
            i += 2
            continue
        if char == "(":
            stack.append(i)
        elif char == ")" and stack:
            start = stack.pop()
            body = pattern[start + 1 : i]
            after = pattern[i + 1] if i + 1 < n else ""
            results.append((body, after))
        i += 1
    return results


def _body_is_repeated_unit(body: str) -> bool:
    """Return ``True`` when a group *body* is itself a repeated/backref unit.

    Such a body, when the enclosing group is also quantified, forms the classic
    nested-quantifier (``(a+)+``) or quantified-backref-in-group (``(\\1)+``)
    catastrophic shape.
    """
    if not body:
        return False
    # The whole body is a backreference, e.g. ``(\1)+``.
    if _BACKREF.fullmatch(body):
        return True
    last = body[-1]
    # A lazy quantifier (``a+?``) ends in ``?``; the real quantifier is the
    # char before it.
    if last == "?":
        return len(body) >= 2 and body[-2] in "+*}"
    if last not in "+*}":
        return False
    # Guard against an escaped quantifier (``a\*``) being read as a quantifier.
    return not (len(body) >= 2 and body[-2] == "\\")


def has_catastrophic_construct(pattern: str) -> bool:
    """Return ``True`` when *pattern* contains a known catastrophic construct.

    Detects nested quantifiers (a quantified group whose body is itself
    quantified) and quantified backreferences.  See the module docstring for the
    residual-risk caveats — this is a structural denylist, not a proof of
    linear-time matching.
    """
    if _QUANTIFIED_BACKREF.search(pattern):
        return True
    for body, after in _group_bodies_with_following_char(pattern):
        if after in _QUANTIFIER_OPENERS and _body_is_repeated_unit(body):
            return True
    return False

File: packages/provide-uterm-go/mcp/patterns.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package mcp

import "regexp"

// Structural ReDoS guard for untrusted (LLM/user-supplied) regex patterns.
// Port of provide.uterm.ai.patterns.has_catastrophic_construct.
//
// Detects the cheap, well-known catastrophic constructs before compilation:
//   - nested quantifiers — a quantified group whose body is itself quantified
//     (e.g. "(a+)+"), the classic exponential-backtracking shape; and
//   - quantified backreferences — "\1+", "(\1)+" …
//
// This is a structural denylist, not a proof of linear-time matching; see the
// Python module docstring for the residual-risk caveats.

// backrefRE matches a backreference token: \1 .. \99.
var backrefRE = regexp.MustCompile(`^\\[1-9][0-9]?$`)

// quantifiedBackrefRE matches a quantified backreference anywhere in the
// pattern: \1+, \2*, \3{...}.
var quantifiedBackrefRE = regexp.MustCompile(`\\[1-9][0-9]?[+*{]`)

// isQuantifierOpener reports whether c, when it immediately follows a group's
// closing paren, makes that group "repeated".
func isQuantifierOpener(c byte) bool {
	return c == '+' || c == '*' || c == '{'
}

// groupBodyAfter pairs a balanced-group body with the character following its
// closing paren.
type groupBodyAfter struct {
	body  string
	after string
}

// groupBodiesWithFollowingChar returns (body, char_after_close) for every
// balanced "(...)" group. Escaped parens are treated as literals; unbalanced
// closes are ignored (the downstream compiler rejects the pattern as invalid),
// so this never panics on malformed input.
func groupBodiesWithFollowingChar(pattern string) []groupBodyAfter {
	var stack []int
	var results []groupBodyAfter
	i, n := 0, len(pattern)
	for i < n {
		c := pattern[i]
		if c == '\\' {
			// Skip the escaped character so "\(" / "\)" are literals.
			i += 2
			continue
		}
		switch {
		case c == '(':
			stack = append(stack, i)
		case c == ')' && len(stack) > 0:
			start := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			body := pattern[start+1 : i]
			after := ""
			if i+1 < n {
				after = pattern[i+1 : i+2]
			}
			results = append(results, groupBodyAfter{body: body, after: after})
		}
		i++
	}
	return results
}

// bodyIsRepeatedUnit reports whether a group body is itself a repeated/backref
// unit. Such a body, when the enclosing group is also quantified, forms the
// classic nested-quantifier or quantified-backref-in-group catastrophic shape.
func bodyIsRepeatedUnit(body string) bool {
	if body == "" {
		return false
	}
	// The whole body is a backreference, e.g. "(\1)+".
	if backrefRE.MatchString(body) {
		return true
	}
	last := body[len(body)-1]
	// A lazy quantifier ("a+?") ends in "?"; the real quantifier precedes it.
	if last == '?' {
		return len(body) >= 2 && (body[len(body)-2] == '+' || body[len(body)-2] == '*' || body[len(body)-2] == '}')
	}
	if last != '+' && last != '*' && last != '}' {
		return false
	}
	// Guard against an escaped quantifier ("a\*") being read as a quantifier.
	return len(body) < 2 || body[len(body)-2] != '\\'
}

// hasCatastrophicConstruct reports whether pattern contains a known
// catastrophic construct (nested quantifier or quantified backreference).
func hasCatastrophicConstruct(pattern string) bool {
	if quantifiedBackrefRE.MatchString(pattern) {
		return true
	}
	for _, g := range groupBodiesWithFollowingChar(pattern) {
		if g.after != "" && isQuantifierOpener(g.after[0]) && bodyIsRepeatedUnit(g.body) {
			return true
		}
	}
	return false
}
